File: python/runners/text-to-image/text_to_image.py
"""Generate an image from a text prompt with a locally pulled Hugging Face diffusion model.

Compiled into a standalone binary with PyInstaller, so it runs without a Python
installation. Designed to be driven by the PHP wrapper:

  * stdout: one JSON object on success, e.g. {"output": "/abs/path/image.png"}
  * stderr: progress and error messages for humans
  * exit code: 0 success, 1 generation failed, 2 invalid arguments

Input is handed to diffusers as-is: values the model cannot handle (a prompt that
is too long, an unsupported size, an unknown device...) make the run fail.
"""

import argparse
import json
import multiprocessing
import os
import sys
import traceback

EXIT_OK = 0
EXIT_FAILED = 1


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="text-to-image",
        description="Generate an image from a text prompt with a locally pulled diffusion model.",
    )
    parser.add_argument("--model", required=True, help="Directory of a pulled diffusers model (contains model_index.json)")
    parser.add_argument("--prompt", required=True, help="Text describing the image")
    parser.add_argument("--output", required=True, help="Path of the image file to write, e.g. image.png")
    parser.add_argument("--negative-prompt", help="Text describing what the image should not contain")
    parser.add_argument("--steps", type=int, help="Number of inference steps (default: the pipeline's own)")
    parser.add_argument("--guidance", type=float, help="Classifier-free guidance scale (default: the pipeline's own)")
    parser.add_argument("--width", type=int, help="Image width in pixels (default: the model's native size)")
    parser.add_argument("--height", type=int, help="Image height in pixels (default: the model's native size)")
    parser.add_argument("--seed", type=int, help="Random seed, for reproducible images")
    parser.add_argument("--device", help="Torch device, e.g. cpu, cuda, mps (default: the best available)")
    return parser.parse_args(argv)


def log(message: str) -> None:
    print(message, file=sys.stderr, flush=True)


def main(argv: list[str]) -> int:
    args = parse_args(argv)

    # Models are loaded from disk only; never reach out to the Hub.
    os.environ["HF_HUB_OFFLINE"] = "1"

    # Imported after argument parsing so --help and usage errors stay fast.
    import torch
    from diffusers import DiffusionPipeline
    from diffusers.utils import logging as diffusers_logging
    from transformers.utils import logging as transformers_logging

    # Progress bars redraw with carriage returns, which garbles pipes and log files.
    show_progress = sys.stderr.isatty()
    if not show_progress:
        diffusers_logging.disable_progress_bar()
        transformers_logging.disable_progress_bar()

    try:
        device = args.device or best_device(torch)
        dtype = torch.float16 if device.startswith("cuda") else torch.float32

        log(f"[text-to-image] Loading {args.model} on {device}...")
        pipeline = DiffusionPipeline.from_pretrained(args.model, torch_dtype=dtype, local_files_only=True)
        pipeline = pipeline.to(device)
        pipeline.set_progress_bar_config(disable=not show_progress)

        options = {
            "negative_prompt": args.negative_prompt,
            "num_inference_steps": args.steps,
            "guidance_scale": args.guidance,
            "width": args.width,
            "height": args.height,
        }
        options = {name: value for name, value in options.items() if value is not None}
        if args.seed is not None:
            options["generator"] = torch.Generator(device="cpu").manual_seed(args.seed)

        log("[text-to-image] Generating image...")
        image = pipeline(prompt=args.prompt, **options).images[0]

        output = os.path.abspath(args.output)
        os.makedirs(os.path.dirname(output), exist_ok=True)
        image.save(output)
    except Exception as e:
        traceback.print_exc()
        log(f"[text-to-image] Error: {type(e).__name__}: {e}")
        return EXIT_FAILED

    print(json.dumps({"output": output}), flush=True)
    log(f"[text-to-image] Saved {output}.")
    return EXIT_OK


def best_device(torch) -> str:
    if torch.cuda.is_available():
        return "cuda"
    if torch.backends.mps.is_available():
        return "mps"
    return "cpu"


if __name__ == "__main__":
    # torch starts multiprocessing helpers by re-running sys.executable, which in the frozen binary is this
    # script; freeze_support() lets those helper invocations run instead of reaching the argument parser.
    multiprocessing.freeze_support()
    sys.exit(main(sys.argv[1:]))

File: src/Runner/TextToImage.php
<?php

declare(strict_types=1);

namespace PhpLovesAi\Runner;

use PhpLovesAi\Binary\Tool;
use PhpLovesAi\Exception\BinaryNotInstalledException;
use PhpLovesAi\Exception\HomeDirectoryNotFoundException;
use PhpLovesAi\Exception\ModelNotFoundException;
use PhpLovesAi\Exception\RunFailedException;
use PhpLovesAi\Exception\UnsupportedModelException;
use PhpLovesAi\Filesystem\Path;

/**
 * Generates images with diffusion models pulled into the project, through the runner installed by
 * `vendor/bin/loves-ai setup text-to-image`.
 *
 *     $image = (new TextToImage())->generate('stabilityai/sd-turbo', 'a cozy cat', __DIR__ . '/cat.png');
 *
 * Input is passed to the model as-is: values it cannot handle (e.g. a prompt that is too long) make the run
 * fail with a RunFailedException.
 */
final class TextToImage extends BinaryRunner
{
    /** A model the runner can load, e.g. stabilityai/sd-turbo; used in error messages. */
    public const EXAMPLE_MODEL = 'stabilityai/sd-turbo';

    /** Hugging Face models this runner can use. */
    public const COMPATIBLE_MODELS_URL = 'https://huggingface.co/models?pipeline_tag=text-to-image&library=diffusers';

    public static function tool(): Tool
    {
        return Tool::TextToImage;
    }

    /**
     * The runner loads models with diffusers' DiffusionPipeline, which needs the model_index.json describing the
     * pipeline. Repositories without it are typically add-ons (embeddings, LoRAs) or single-file checkpoints.
     */
    protected function ensureModelIsSupported(string $model, string $modelPath): void
    {
        if (is_file("{$modelPath}/model_index.json")) {
            return;
        }

        throw new UnsupportedModelException($model, $modelPath, sprintf(
            '%s cannot generate images: it is not a complete Diffusers text-to-image model (%s/model_index.json is missing). '
            . 'It may be an add-on such as embeddings or a LoRA, which only works on top of a base model, a model in another '
            . 'format, or a pull that did not finish. Use a Diffusers text-to-image model instead, e.g. %s (browse: %s).',
            $model,
            $modelPath,
            self::EXAMPLE_MODEL,
            self::COMPATIBLE_MODELS_URL,
        ));
    }

    /**
     * @param string                        $model          Hugging Face model id, e.g. "stabilityai/sd-turbo"
     * @param string                        $prompt         text describing the image
     * @param string                        $outputPath     image file to write, e.g. "cat.png"; its extension picks
     *                                                      the format. A leading "~" is expanded
     * @param string|null                   $negativePrompt text describing what the image should not contain
     * @param int|null                      $steps          inference steps; null uses the pipeline's default
     * @param float|null                    $guidanceScale  classifier-free guidance; null uses the pipeline's default
     * @param int|null                      $width          image width in pixels; null uses the model's native size
     * @param int|null                      $height         image height in pixels; null uses the model's native size
     * @param int|null                      $seed           random seed, for reproducible images
     * @param string|null                   $device         torch device such as "cpu", "cuda" or "mps";
     *                                                      null picks the best available
     * @param (callable(string): void)|null $onOutput       receives the runner's progress output as it arrives
     *
     * @return string absolute path of the generated image
     *
     * @throws BinaryNotInstalledException
     * @throws ModelNotFoundException
     * @throws UnsupportedModelException when the model is not a Diffusers text-to-image model
     * @throws HomeDirectoryNotFoundException
     * @throws RunFailedException
     */
    public function generate(
        string $model,
        string $prompt,
        string $outputPath,
        ?string $negativePrompt = null,
        ?int $steps = null,
        ?float $guidanceScale = null,
        ?int $width = null,
        ?int $height = null,
        ?int $seed = null,
        ?string $device = null,
        ?callable $onOutput = null,
    ): string {
        $result = $this->run($model, [
            '--prompt' => $prompt,
            '--output' => Path::expandHome($outputPath),
            '--negative-prompt' => $negativePrompt,
            '--steps' => $steps,
            '--guidance' => $guidanceScale,
            '--width' => $width,
            '--height' => $height,
            '--seed' => $seed,
            '--device' => $device,
        ], $onOutput);

        if (!is_string($result['output'] ?? null)) {
            throw new RunFailedException(0, 'The runner did not report the path of the generated image.');
        }

        return $result['output'];
    }
}
